RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+let mutable s = "hello"
+
+// s is now (s | null), despite the warning
+s <- null
+```
+
+This includes function composition:
+
+```fsharp
+// val f1 : 'a -> 'a
+let f1 s = s
+
+// val f2 : string -> (string | null)
+let f2 s = if s <> "" then "hello" else null
+
+// val f3 : string -> string
+let f3 (s: string) =
+    match s with
+    | null -> ""
+    | _ -> s
+
+// val pipeline : string -> (string | null)
+let pipeline = f1 >> f2 >> f3 // WARNING - flowing possible null as input to function that assumes a non-null input
+```
+
+Existing nullability rules for F# types hold (i.e., you cannot suddenly assign `null` to an F# record).
+
+### Tooling considerations
+
+To remain in line our first principle:
+
+> On the surface, F# is "almost" as simple to use as it is today. In practice, it must feel simpler due to nullability of types like `string` being explicit rather than implicit.
+
+We'll consider careful suppression of nullability in F# tooling so that the extra information doesn't appear overwhelming. For example, imagine the following signature in tooltips:
+
+```fsharp
+member Join : source2: (Expression<seq<string | null> | null> | null) * key1:(Expression<Func<(string | null), T | null> | null> | null)) * key2:(Expression<Func<(string | null), T | null> | null> | null))
+```
+
+AGH! This signature is challenging enough already, but now nullability has made it significantly worse.
+
+#### F# tooltips
+
+In QuickInfo tooltips today, we reduce the length of generic signatures with a section below the signature. We can do a similar thing for saying if something is nullable:
+
+```
+member Foo : source: (seq<'T>) * predicate: ('T -> 'U- > bool) * item: 'U
+
+    where 'T, 'U are nullable
+
+Generic parameters:
+'T is string
+'U is int
+```
+
+Although this uses more vertical space in the tooltip, it reduces the length of the signature, which can already difficult to parse for many members in .NET.
+
+In Signature Help tooltips today, we embed the concrete type for a generic type parameter in the tooltip. We may consider doing a similar thing as with QuickInfo, where nullability is expressed separately from the signature. However, this would also likely require a change where we factor out concrete types from the type signature at the parameter site. For example:
+
+```
+List(System.Collections.Generics.IEnumerable<'T>)
+
+    where 'T is (string | null)
+
+Initializes an instance of a [...]
+```
+
+#### Signatures in F# Interactive
+
+To have parity with F# signatures (note: not tooltips), signatures printed in F# interactive will have nullability expression as if it were in code:
+
+```
+> let s: string | null = "hello";;
+
+    val s : string | null = "hello"
+```
+
+#### F# scripting
+
+F# scripts will now also give warnings when using the compiler that implements this feature. This also necessitates a new feature for F# scripting that allows you to opt-out of any nullability warnings. Example:
+
+```fsharp
+#nullability "false"
```

Comment:
Probably just the existing

    #nowarn "nullness"

where "nullness" is a package of nowarns